### PR TITLE
AArch64: Implement buildArgs() and buildDirectDispatch() in PrivateLinkage

### PR DESCRIPTION
**Pull request by knn-k**
This commit implements buildArgs(), buildDirectDispatch(), etc. for
ARM64PrivateLinkage.

Signed-off-by: KONNO Kazuhiro <konno@jp.ibm.com>

### CODE REVIEW

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by knn-k** [comment]
This PR is still WIP.

**Comment by knn-k** [comment]
I will work on `buildVirtualDispatch()` in another PR.

`x19` (vmThread) and `x20` (Java SP) has to be marked as `Preserved` here in the same way as p and arm32.  Otherwise the register allocation fails.

**Comment by knn-k** [comment]
Moved the implementation of `buildIndirectDispatch()` to #6761.

**Comment by knn-k** [comment]
@0xdaryl Should I separate `buildDirectDispatch()` implementation and open another PR for that, so that this PR focuses on `buildArgs()`?

#6867 depends on `buildPrivateLinkageArgs()` added by this PR.

**Comment by knn-k** [comment]
I opened #6877 and moved the declaration of `TR::ARM64PrivateLinkage::buildPrivateLinkageArgs()` to it, because #6867 depends on the function.

**Comment by knn-k** [comment]
I rebased and resolved the merge conflict with #6877.

**Comment by knn-k** [comment]
Fixed the offsets of the arguments passed via Java stack.

**Comment by knn-k** [comment]
Resolved a merge conflict with #6689.

**Comment by knn-k** [comment]
Resolved a merge conflict.

**Comment by knn-k** [comment]
Fixed the register dependency for floating-point return value.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 1]
```diff
@@ -293,14 +299,396 @@ int32_t TR::ARM64PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
...
+            else // numIntegerArgs >= numIntArgRegs
+               {
+               if (childType == TR::Address)
+                  {
+                  op = TR::InstOpCode::strpostx;
+                  }
+               else
+                  {
+                  op = TR::InstOpCode::strpostw;
+                  }
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: More succinctly as:
```
   op = (childType == TR::Address) ? TR::InstOpCode::strpostx : TR::InstOpCode::strpostw;
```

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 2]
```diff
@@ -293,14 +299,396 @@ int32_t TR::ARM64PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
...
+                  dependencies->addPostCondition(resultReg, TR::RealRegister::v0);
+                  }
+               else
+                  TR::addDependency(dependencies, argRegister, properties.getFloatArgumentRegister(numFloatArgs), TR_FPR, cg());
+               }
+            else // numFloatArgs >= numFloatArgRegs
+               {
+               mref = getOutgoingArgumentMemRef(argMemReg, argRegister, TR::InstOpCode::vstrpostd, pushToMemory[argIndex++]);
+               }
+            numFloatArgs++;
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: The double case is nearly identical to the float case except for the `pushFloatArg()/pushDoubleArg()` call, and the two opcodes for the generated instructions.  Can you try and combine these cases in some way to reduce the code size?

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 3]
```diff
@@ -293,14 +299,396 @@ int32_t TR::ARM64PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
...
+                  }
+               else
+                  {
+                  op = TR::InstOpCode::strpostw;
+                  }
+               mref = getOutgoingArgumentMemRef(argMemReg, argRegister, op, pushToMemory[argIndex++]);
+               }
+            numIntegerArgs++;
+            break;
+         case TR::Int64:
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: Why can't the `TR::Int64` case be combined with the above case?  It looks nearly identical.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 4]
```diff
@@ -293,14 +299,396 @@ int32_t TR::ARM64PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
...
+         specialArgReg = getProperties().getVTableIndexArgumentRegister();
+         break;
+      case TR::java_lang_invoke_MethodHandle_invokeWithArgumentsHelper:
+         numIntArgRegs   = 0;
+         numFloatArgRegs = 0;
+         break;
+      }
+   if (specialArgReg != TR::RealRegister::NoReg)
+      {
+      TR_UNIMPLEMENTED();
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: I opened Issue #7023 to track it.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 5]
```diff
@@ -193,14 +202,406 @@ void TR::ARM64PrivateLinkage::createEpilogue(TR::Instruction *cursor)
...
+            if (numIntegerArgs >= numIntArgRegs)
+               numMemArgs++;
+            numIntegerArgs++;
+            totalSize += TR::Compiler->om.sizeofReferenceAddress();
+            break;
+         case TR::Int64:
+            if (numIntegerArgs >= numIntArgRegs)
+               numMemArgs++;
+            numIntegerArgs++;
+            totalSize += 2*TR::Compiler->om.sizeofReferenceAddress();
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: Merged the `TR::Int64` case with other integer cases into one.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 6]
```diff
@@ -193,14 +202,406 @@ void TR::ARM64PrivateLinkage::createEpilogue(TR::Instruction *cursor)
...
+            break;
+         case TR::Float:
+            if (numFloatArgs >= numFloatArgRegs)
+               numMemArgs++;
+            numFloatArgs++;
+            totalSize += TR::Compiler->om.sizeofReferenceAddress();
+            break;
+         case TR::Double:
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: Merged the `TR::Double` case with the ` `TR::Float` case.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp (hunk 1 above)
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: Changed to use the `? :` operator.
Also adding the `TR::Int64` case.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp (hunk 3 above)
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: Merged the `TR::Int64` case with other integer cases.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp (hunk 2 above)
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: Merged the `TR::Double` case with the `TR::Float` case.

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp [hunk 7]
```diff
@@ -293,14 +299,396 @@ int32_t TR::ARM64PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
...
 
 TR::Register *TR::ARM64PrivateLinkage::buildDirectDispatch(TR::Node *callNode)
    {
-   TR_UNIMPLEMENTED();
-   return NULL;
+   TR::SymbolReference *callSymRef = callNode->getSymbolReference();
+   const TR::ARM64LinkageProperties &pp = getProperties();
+   TR::RegisterDependencyConditions *dependencies =
+      new (trHeapMemory()) TR::RegisterDependencyConditions(
+         pp.getNumberOfDependencyGPRegisters(),
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: `TR::PPCPrivateLinkage::buildDirectDispatch()` does the same for allocating `RegisterDependencyConditions` as shown below.  Is the p version unsafe, too?

https://github.com/eclipse/openj9/blob/b2e5ef80d9577310a0dc0bdfcf096196845407fa/runtime/compiler/p/codegen/PPCPrivateLinkage.cpp#L2721-L2724

[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: `_numberOfDependencyGPRegisters` is defined as follows in p, including the number of FP registers.  I think I have to change the definition of `_numberOfDependencyGPRegisters` for aarch64.

https://github.com/eclipse/openj9/blob/b2e5ef80d9577310a0dc0bdfcf096196845407fa/runtime/compiler/p/codegen/PPCPrivateLinkage.cpp#L329-L330

[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: Updated `_numberOfDependencyGPRegisters`.

**Comment by knn-k** [comment]
The Travis CI build failed with the following error:
> fatal: unable to access 'https://github.com/ibmruntimes/openj9-openjdk-jdk11.git/': Failed to connect to github.com port 443: Connection timed out

File: runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (0xdaryl) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: If you're going to fix it that way then the name of this API should change as it is not just accounting for GP registers any more.  At the very least, drop the "GP" from the name.

[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp]: The API and the property member are in the OMR side.  Please let me do that later.  I will open an issue to track it.

**Comment by knn-k** [comment]
Rebased to kick the Travis CI build.

**Comment by 0xdaryl** [comment]
Jenkins test sanity alinux64xl jdk11

**Pull request merged**